packages/blocks/shuffle/src/components/add-entities-dialog.tsx

```diff
@@ -0,0 +1,220 @@
+
+const _AddEntitiesDialog: ForwardRefRenderFunction<
+  AddEntitiesDialogRef,
+  AddEntitiesDialogProps
+> = ({ entityTypes, blockEntityId, onAddEntityItems, graphService }, ref) => {
+  const [dialogState, setDialogState] = useState<DialogState>("closed");
+  const [entityList, setEntityList] = useState<Entity[]>([]);
+  // `selections` is a object that stores checkbox `checked` state for each entity.
+  const [selections, setSelections] = useState<Record<string, boolean>>({});
+
+  /**
+   * Defined a custom ref, so the parent can simply call `dialogRef.show()`,
+   * which also resets the state to it's initial value when modal is opened,
+   * instead of using an `useEffect` that listens modal's state
+   */
+  useImperativeHandle(
+    ref,
+    () => ({
+      show: () => {
+        setDialogState("entityTypes");
+        setEntityList([]);
+      },
+    }),
+    [],
+  );
+
+  const handleEntityTypeClick = async (entityTypeId: string) => {
+    if (!graphService) return;
+
+    // get the entities of clicked entity type
+    const { data } = await graphService.aggregateEntities({
+      data: { operation: { entityTypeId, itemsPerPage: 100 } },
+    });
+
+    if (!data) return setDialogState("closed");
+
+    const { results: entities } = data;
+
+    // by default, all entity checkboxes are checked
+    const initialEntitySelection = entities.reduce(
+      (prev, curr) => ({ ...prev, [curr.entityId]: true }),
+      {},
+    );
+
+    setEntityList(entities);
+    setSelections(initialEntitySelection);
+    setDialogState("entities");
+  };
+
+  const handleCheckboxChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setSelections((prev) => ({
+      ...prev,
+      [event.target.name]: event.target.checked,
+    }));
+  };
+
+  const handleClose = () => setDialogState("closed");
+
+  const handleAddClick = async () => {
+    if (!graphService) return;
+
+    // create links for selected entities
+    const selectedEntityIds = Object.entries(selections)
+      .filter(([_, isSelected]) => isSelected)
+      .map(([id]) => id);
+
+    const createLinkPromises = selectedEntityIds.map((entityId) =>
+      graphService.createLink({
+        data: {
+          sourceEntityId: blockEntityId,
+          destinationEntityId: entityId,
+          path: "$.items",
+        },
+      }),
+    );
+
+    const createLinkResponses = await Promise.all(createLinkPromises);
+
+    // add the new items to the shuffle items list
+    onAddEntityItems(
+      selectedEntityIds.map((entityId, i) => {
+        return {
+          id: uuid(),
+          value: "",
+          entityId,
+          linkId: createLinkResponses[i]?.data?.linkId,
+        };
+      }),
+    );
+
+    // close the modal
+    handleClose();
+  };
+
+  const selectedItemCount = useMemo(() => {
+    const entries = Object.entries(selections);
+    const selectedEntries = entries.filter(([_, isSelected]) => isSelected);
+    return selectedEntries.length;
```

Comment:
By turning selections into an array we could simplify this.
```suggestion
    return selections.length;
```